stdlib/Printf/src/Printf.jl

```diff
@@ -2,1297 +2,770 @@
+end
+
+@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Chars}
+    leftalign, width = spec.leftalign, spec.width
+    if !leftalign && width > 1
+        for _ = 1:(width - 1)
+            buf[pos] = UInt8(' ')
+            pos += 1
+        end
+    end
+    pos = writechar(buf, pos, arg isa String ? arg[1] : Char(arg))
+    if leftalign && width > 1
+        for _ = 1:(width - 1)
+            buf[pos] = UInt8(' ')
+            pos += 1
         end
     end
+    return pos
 end
 
-# returns the number of (ASCII) chars output by print_fixed
-function print_fixed_width(precision, pt, ndigits, trailingzeros=true)
-    count = 0
-    if pt <= 0
-        # 0.0dddd0
-        count += 2
-        precision += pt
-        if pt < 0
-            count -= pt
+# strings
+@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Strings}
+    leftalign, hash, width, prec = spec.leftalign, spec.hash, spec.width, spec.precision
+    str = string(arg)
+    op = p = prec == -1 ? (length(str) + (hash ? arg isa AbstractString ? 2 : 1 : 0)) : prec
+    if !leftalign && width > p
+        for _ = 1:(width - p)
+            buf[pos] = UInt8(' ')
+            pos += 1
         end
-        count += ndigits
-        precision -= ndigits
-    elseif ndigits <= pt
-        # dddd000.000000
-        count += ndigits
-        if ndigits < pt
-            count += pt - ndigits
+    end
+    if hash
+        if arg isa Symbol
+            buf[pos] = UInt8(':')
+            pos += 1
+            p -= 1
+        elseif arg isa AbstractString
+            buf[pos] = UInt8('"')
+            pos += 1
+            p -= 1
         end
-        count += trailingzeros
-    else # 0 < pt < ndigits
-        # dd.dd0000
-        ndigits -= pt
-        count += pt + 1 + ndigits
-        precision -= ndigits
     end
-    if trailingzeros && precision > 0
-        count += precision
+    for c in str
+        p == 0 && break
+        pos = writechar(buf, pos, c)
+        p -= 1
+    end
+    if hash && arg isa AbstractString && p > 0
+        buf[pos] = UInt8('"')
+        pos += 1
+    end
+    if leftalign && width > op
+        for _ = 1:(width - op)
+            buf[pos] = UInt8(' ')
+            pos += 1
+        end
     end
-    return count
+    return pos
 end
 
-# note: if print_fixed is changed, print_fixed_width should be changed accordingly
-function print_fixed(out, precision, pt, ndigits, trailingzeros=true, buf = Grisu.getbuf())
-    pdigits = pointer(buf)
-    if pt <= 0
-        # 0.0dddd0
-        print(out, '0')
-        print(out, '.')
-        precision += pt
-        while pt < 0
-            print(out, '0')
-            pt += 1
+# integers
+@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Ints}
+    leftalign, plus, space, zero, hash, width, prec =
+        spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
+    bs = base(T)
+    arg2 = arg isa AbstractFloat ? Integer(trunc(arg)) : arg
```

Comment:
`AbstractFloat` and `Integer` aren't the only type classes that can exist. Other user types may need the chance to override this too.

Reply:
You'll have to be more concrete with what you'd like changed here, since it's not very clear from your comment. If were to guess, you're imagining we'd want something similar to `Printf.tofloat` for integers, so having a `Printf.tointeger` that custom types could overload if they'd like a custom type formatted as an integer. I thought of that, but the interface for an "integer" is also much simpler than floats; i.e. if your custom type supports multiplication, division, and `ndigits`, then it'll "just work". This is one of those cases where if someone has a valid use-case, it'd be pretty easy to add, but no use pre-engineering if not necessary.

Reply:
Yeah, that might help? Currently this only handles half of the cases, just in Base:
```
julia> subtypes(Real)
4-element Vector{Any}:
 AbstractFloat
 AbstractIrrational
 Integer
 Rational
```
Which doesn't always work now:
```
julia> @printf "%d" π
ERROR: MethodError: no method matching Integer(::Irrational{:π})
julia> @printf "%d" 2//4
ERROR: InexactError: Integer(1//2)
```
But sometimes it does:
```
julia> @printf "%d" 2//1
2
```